### PR TITLE
Add support for tiled indices via padding/alignment aware embedding kernel (tiled indices only)

### PR DESCRIPTION
**Pull request by yugi957**
### Ticket
[#14915](https://github.com/tenstorrent/tt-metal/issues/14915)

### Problem description
uint32 untilize not accurate, so padding aware kernel was needed as opposed to standard conversion

### What's changed
A lot of changes to look at, first is the split_work_to_cores_aligned function.

I made this function to help split work to cores such that the units per core is aligned with some value. In the new embedding kernel, I align with the face_height of a tile. After developing this kernel further, I believe it's possible to use the standard split_work function for more parallelism, but I figured I'd keep it for now as it could prove useful for perf in the future where we want to make sure our split work is aligned. (You can use the new aligned function as an immediate drop in replacement for the same return values.)

As for the new program factory and kernel, the principle of weights writing is identical to the RM kernel, and this means we want to read the indices such that we are accessing the tensor in an RM fashion, you could compare it simply to a parallelized print. A lot of the work is simply offset algebra, and I tried to make sure to use const expr for any tile or face heights, do let me know if there's anywhere I don't!

With this change, sweeps for embedding go from 46% -> 92% due to the support of tiled indices.

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/12641263680)
- [x] [Model regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/12589016962)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+    constexpr uint32_t row_length = get_compile_time_arg_val(8);
+    constexpr uint32_t input_block_size_bytes = get_compile_time_arg_val(9);
+
+    constexpr uint32_t cb_id_in0 = 0;
+    constexpr uint32_t cb_id_in1 = 1;
+    constexpr uint32_t cb_id_in2 = 2;
+
+    constexpr uint32_t face_size = 16;
+    constexpr uint32_t tile_height = 32;
```
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:54]: Can we send this in as a param from host using tile shape. Look up how to get the tile height from the tile shape from the tensor.

Tldr, in the future we are going to support tiles that are not 32x32, so we want to get that from the tile shape api.

Comment (ntarafdar) [reply ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:54]: When I say this, I mean the tile height

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+#endif
+    // bool printed = false;
+    auto read_block = [&](const uint32_t& token_idx, const uint32_t& width_size, const uint32_t& offset = 0) {
+        cb_reserve_back(cb_id_in0, 1);
+        uint32_t weight_l1_addr = get_write_ptr(cb_id_in0);
+        volatile tt_l1_ptr uint16_t* weight_l1_ptr = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(weight_l1_addr);
+        uint64_t src_noc_addr;
+        uint32_t token = input_l1_ptr[token_idx + offset];
+
+        // I would like to keep this for future debugging purposes
```
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:102]: If we want to keep this around, let's add comments around the magic numbers, like what's 105?

**Review by ntarafdar — APPROVED** [review]
Amazing work, just some small comments regarding using the API to get tile height rather than assuming 32 tile height and commenting on magic numbers even if we were to leave commented out debug code!

**Comment by tt-aho** [comment]
Looks like this code was forked/copied before my recent pr https://github.com/tenstorrent/tt-metal/pull/16237 was merged.

Would be good to copy over the cleanup/refactoring, especially for the kernel side.

File: tt_metal/common/work_split.cpp [hunk 3]
```diff
@@ -418,5 +418,47 @@ std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_
...
+    // Core set for all active cores
+    CoreRangeSet all_cores = num_cores_to_corerangeset(required_cores, grid_size, false);
+
+    // Calculate remaining units for the last core
+    uint32_t evenly_distributed_units = (required_cores - 1) * units_per_core;
+    uint32_t remaining_units = units_to_divide - evenly_distributed_units;
+
+    // Create core groups
+    CoreRangeSet core_group_1 = all_cores;
+    CoreRangeSet core_group_2;
```
Comment (sjameelTT) [inline tt_metal/common/work_split.cpp:447]: default initialization is empty?

File: tt_metal/common/work_split.hpp [hunk 4]
```diff
@@ -53,8 +53,8 @@ CoreRangeSet num_cores_to_corerangeset_in_subcoregrids(
 std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores(
     const CoreCoord grid_size, const uint32_t units_to_divide, const bool row_wise = false);
 
-std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores(
-    const CoreRangeSet& core_grid, const uint32_t units_to_divide, const bool row_wise = false);
+std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores_aligned(
+    const CoreCoord grid_size, const uint32_t units_to_divide, const uint32_t alignment);
```
Comment (sjameelTT) [inline tt_metal/common/work_split.hpp:57]: const reference CoreCoord?

File: tt_metal/common/work_split.cpp [hunk 5]
```diff
@@ -418,5 +418,47 @@ std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_
         target_num_cores, all_cores, core_group_1, core_group_2, units_per_core_group_1, units_per_core_group_2);
 }
 
+std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores_aligned(
+    const CoreCoord grid_size, const uint32_t units_to_divide, const uint32_t alignment) {
```
Comment (sjameelTT) [inline tt_metal/common/work_split.cpp:422]: const reference for CoreCoord?

can the existing work split this if the units of work = align(tensor_side, alignment)? Mostly just curious about whether or not you could have used the existing one.

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp [hunk 6]
```diff
@@ -22,8 +22,8 @@ void Embeddings::validate(const std::vector<Tensor> &input_tensors) const {
     TT_FATAL(input_tensors.size() == 2, "Must have between 2 input tensors");
     auto &a = input_tensors.at(0);
     const auto &weights = input_tensors.at(1);
-    TT_FATAL(a.get_layout() == Layout::ROW_MAJOR, "Input tensor must be Row Major Layout");
-    TT_FATAL(weights.get_layout() == Layout::ROW_MAJOR, "Weights tensor must be Row Major Layout");
+    // TT_FATAL(a.get_layout() == Layout::ROW_MAJOR, "Error");
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:25]: delete

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp [hunk 7]
```diff
@@ -22,8 +22,8 @@ void Embeddings::validate(const std::vector<Tensor> &input_tensors) const {
...
+    TT_FATAL(weights.get_layout() == Layout::ROW_MAJOR, "Error");
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:26]: more descriptive error message

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp [hunk 8]
```diff
@@ -45,7 +45,7 @@ void Embeddings::validate(const std::vector<Tensor> &input_tensors) const {
         TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Embedding only supports interleaved RM outputs");
         TT_FATAL(!is_block_float(this->output_dtype), "Output cannot be a block float dtype when not tilized");
     }
-    TT_FATAL(a.get_legacy_shape()[1] == 1 && a.get_legacy_shape()[2] == 1, "Only dim 0 && 3 for the input can be non 1");
+    // TT_FATAL(a.get_legacy_shape()[1] == 1 && a.get_legacy_shape()[2] == 1, "Only dim 0 && 3 for the input can be non 1");
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:48]: delete

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp [hunk 9]
```diff
@@ -56,9 +56,9 @@ void Embeddings::validate(const std::vector<Tensor> &input_tensors) const {
 std::vector<TensorSpec> Embeddings::compute_output_specs(const std::vector<Tensor> &input_tensors) const {
     const auto &input_tensor = input_tensors.at(0);
     const auto &weight_tensor = input_tensors.at(1);
-    auto num_output_embeddings = input_tensor.logical_shape()[3];
+    auto num_output_embeddings = input_tensor.logical_shape()[-1];
     auto batch_num = input_tensor.logical_shape()[0];
-    auto num_embedding_dims = weight_tensor.logical_shape()[3];
+    auto num_embedding_dims = weight_tensor.logical_shape()[-1];
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:61]: This was from a previous change but careful when using `logical_shape`.  

`logical_shape()` is the async getter, which is fairly untested and `get_logical_shape()` is the standard synchronous one. Clearly seems like it hasn't caused bugs here but it could be problematic.

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        // This prints which token is being read, as well as one full tile per core
+        // DPRINT << "Token: " << token << ENDL();
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:104]: Unrelated to the PR. We need a standardized environment variable for kernel debugging statements going forward if we're going to keep them.

**Review by sjameelTT — COMMENTED** [review]
Good stuff, ping me when the comments are addressed

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp (hunk 1 above)
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:54]: Also, we are moving towards passing in cb id as compile-time args instead of hard-coding.

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp (hunk 10 above)
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:104]: Why not use `TileSlice` and `SliceRange`? This seems like implementing existing debug tools

File: ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp (hunk 1 above)
Comment (tt-aho) [reply ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embedding_ind_tilized.cpp:54]: removing hardcoded cb ids is related to the cleanup requested here https://github.com/tenstorrent/tt-metal/pull/16296#issuecomment-2560374212

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 11]
```diff
@@ -533,15 +535,255 @@ operation::ProgramWithCallbacks embeddings_rm(
...
+    // weights shape is [1, 1, num_embeddings, num_dim]
+    uint32_t num_embeddings = weights.get_legacy_shape()[-2];
+
+    uint32_t batch_size = a.get_logical_shape()[0];  // num rows
+    uint32_t num_cols = a.get_logical_shape()[-1];
+    uint32_t volume = num_cols * batch_size;
+    constexpr uint32_t alignment = 32;
+    constexpr uint32_t tile_size = 32;
+    constexpr uint32_t face_size = 16;
+    constexpr uint32_t face_volume = face_size * face_size;
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:586]: These shouldn't be hardcoded here. These constants are in `tt_metal/common/constants.hpp`

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 12]
```diff
@@ -533,15 +535,255 @@ operation::ProgramWithCallbacks embeddings_rm(
...
+                runtime_args[0] = input_dram_buffer->address();
+                runtime_args[1] = weights_dram_buffer->address();
+            }
+
+            {
+                auto& runtime_args = GetRuntimeArgs(program, writer_kernel_id, core);
+                runtime_args[0] = output_dram_buffer->address();
+            }
+        }
+    };
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:771]: Should refactor how `GetRuntimeArgs` is called similar to the changes from my pr for the two other functions

**Review by tt-aho — CHANGES_REQUESTED** [review]
> Looks like this code was forked/copied before my recent pr #16237 was merged.
> 
> Would be good to copy over the cleanup/refactoring, especially for the kernel side.

Please address this before merge.

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp (hunk 7 above)
Comment (tt-aho) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:25]: Seems like a merge conflict should revert the message change

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp (hunk 8 above)
Comment (tt-aho) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_device_operation.cpp:47]: This seems like this assert should still be here if a is RM, and a new assert for a's shape should be added when tilized

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 13]
```diff
@@ -4,16 +4,64 @@
...
+    uint32_t total_cores = num_cores_x * num_cores_y;
+
+    // Initialize units_per_core and required_cores
+    uint32_t units_per_core = alignment;
+    uint32_t required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
+
+    // Double units_per_core until it fits within total_cores
+    while (required_cores > total_cores) {
+        units_per_core += alignment;
+        required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:30]: Minor: Can use div_up

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 14]
```diff
@@ -4,16 +4,64 @@
...
+std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores_aligned(
+    const CoreCoord grid_size, const uint32_t units_to_divide, const uint32_t alignment) {
+    ZoneScoped;
+
+    uint32_t num_cores_x = grid_size.x, num_cores_y = grid_size.y;
+    uint32_t total_cores = num_cores_x * num_cores_y;
+
+    // Initialize units_per_core and required_cores
+    uint32_t units_per_core = alignment;
+    uint32_t required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:25]: Minor: Can use div_up

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 15]
```diff
@@ -4,16 +4,64 @@
...
+
+    // Double units_per_core until it fits within total_cores
+    while (required_cores > total_cores) {
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:28]: Comment seems slightly incorrect. It only really doubles on the first iteration.

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 16]
```diff
@@ -533,15 +581,267 @@ operation::ProgramWithCallbacks embeddings_rm(
...
+        (std::uint32_t)src0_cb_index,
+        (std::uint32_t)src1_cb_index,
+        (std::uint32_t)src2_cb_index,
+        (std::uint32_t)in0_is_dram,
+        (std::uint32_t)input_page_size,
+        (std::uint32_t)weights_is_dram,
+        (std::uint32_t)weight_page_size,
+        (std::uint32_t)block_height,
+        (std::uint32_t)block_height * input_element_size_bytes};
+    */
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:714]: Remove?

**Review by tt-aho — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp (hunk 15 above)
Comment (yugi957) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:28]: Yup gotta fix that

File: tests/ttnn/unit_tests/operations/test_embedding.py [hunk 17]
```diff
@@ -131,8 +131,117 @@ def test_moe_embedding(
...
+    output_mem_config,
+    indices_layout,
+    weight_layout,
+    output_layout,
+):
+    torch.manual_seed(1234)
+
+    torch_input_tensor = torch.randint(0, vocabulary_size - 1, (batch_size, sentence_size))
+    torch_weights = torch_random((vocabulary_size, hidden_embedding_dim), -0.1, 0.1, dtype=torch.bfloat16)
+    # torch_output_tensor = torch.nn.functional.embedding(torch_input_tensor, torch_weights)
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_embedding.py:153]: remove

File: tests/ttnn/unit_tests/operations/test_embedding.py [hunk 18]
```diff
@@ -146,17 +255,20 @@ def test_embedding_tiled_input(
 
     torch_input_tensor = torch.randint(0, vocabulary_size - 1, (batch_size, sentence_size))
     torch_weights = torch_random((vocabulary_size, hidden_embedding_dim), -0.1, 0.1, dtype=torch.bfloat16)
+    # row_indices = torch.arange(0, vocabulary_size, dtype=torch.float32)
+    # torch_weights = row_indices.unsqueeze(1).expand(-1, hidden_embedding_dim)
+    # print(torch_weights)
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_embedding.py:260]: remove

File: tests/ttnn/unit_tests/operations/test_embedding.py [hunk 19]
```diff
@@ -173,6 +285,11 @@ def test_embedding_tiled_input(
     )
     output_tensor = ttnn.to_torch(output_tensor)
 
+    # print("Reversed ttnn Tensor:")
+    # print(reverse_embedding_output(output_tensor, torch_weights))
+    # print("Reversed torch Tensor:")
+    # print(reverse_embedding_output(torch_output_tensor, torch_weights))
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_embedding.py:291]: remove

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 20]
```diff
@@ -4,16 +4,64 @@
...
+#include "ttnn/operations/core/core.hpp"
+#include "tt_metal/host_api.hpp"
 #include "tt_metal/common/constants.hpp"
 #include "tt_metal/detail/util.hpp"
 #include "tt_metal/host_api.hpp"
 #include "tt_log.h"
 
 using namespace tt;
 
+std::tuple<uint32_t, CoreRangeSet, CoreRangeSet, CoreRangeSet, uint32_t, uint32_t> split_work_to_cores_aligned(
```
[the remark this answers is not in the document]
Comment (yugi957) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp]: edited, does this look better?

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 21]
```diff
@@ -4,16 +4,73 @@
...
+    uint32_t total_cores = num_cores_x * num_cores_y;
+
+    // Initialize units_per_core and required_cores
+    uint32_t units_per_core = alignment;
+    uint32_t required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
+
+    // += alignment until it fits within total_cores
+    while (required_cores > total_cores) {
+        units_per_core += alignment;
+        required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:39]: should we just do smth like instead of looping or it won't suffice?

```
    if (required_cores > total_cores) {
        units_per_core = ((units_to_divide + total_cores - 1) / total_cores + alignment - 1) / alignment * alignment;
        required_cores = (units_to_divide + units_per_core - 1) / units_per_core;
    }
```

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 22]
```diff
@@ -4,16 +4,73 @@
...
+struct CoreSplitResult {
+    uint32_t required_cores = 0;
+    CoreRangeSet all_cores;
+    CoreRangeSet core_group_1;
+    CoreRangeSet core_group_2;
+    uint32_t units_per_core_group_1 = 0;
+    uint32_t units_per_core_group_2 = 0;
+};
+
+CoreSplitResult split_work_to_cores_aligned(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:25]: please move to .cpp

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp [hunk 23]
```diff
@@ -533,15 +590,261 @@ operation::ProgramWithCallbacks embeddings_rm(
     return {.program = std::move(program), .override_runtime_arguments_callback = override_runtime_arguments_callback};
 }
 
+operation::ProgramWithCallbacks embeddings_tilized_indices(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:593]: can you please move all implementations to .cpp?

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp (hunk 21 above)
Comment (yugi957) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:39]: Yes I think this is similar to the implementation I was looking for earlier before settling on brute force for the moment. Let me try it

File: ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp (hunk 23 above)
Comment (yugi957) [reply ttnn/cpp/ttnn/operations/embedding/device/embedding_program_factory.hpp:593]: Will do in a separate PR: https://github.com/tenstorrent/tt-metal/issues/16477
